File: api/src/main/java/io/github/mh321productions/serverapi/module/log/LogModule.java
package io.github.mh321productions.serverapi.module.log;

import java.util.ArrayList;

import javax.annotation.Nullable;

import io.github.mh321productions.serverapi.Main;
import io.github.mh321productions.serverapi.api.APIImplementation;
import io.github.mh321productions.serverapi.api.SubPlugin;
import org.bukkit.plugin.Plugin;

import io.github.mh321productions.serverapi.module.Module;
import io.github.mh321productions.serverapi.module.ModuleStopFunction;
import io.github.mh321productions.serverapi.module.ModuleType;
import org.jetbrains.annotations.NotNull;

/**
 * Das Logging-Modul erstellt für jedes registrierte Sub-Plugin ein eigenes <br/>
 * Logfile, welches automatisch über den internen Logger <br/>
 * ({@link Plugin#getLogger()}) beschrieben wird. Wenn das Modul gestoppt <br/>
 * oder das Sub-Plugin entladen wird, wird das Logfile geschlossen. <br/>
 * Das Logfile ist unter "plugins/ServerAPI/logs/&lt;Sub-Plugin Name&gt;.txt" <br/>
 * zu finden.
 * @author 321Productions
 */
public class LogModule extends Module {
	
	private ArrayList<LogWrapper> logWrappers = new ArrayList<>();

	public LogModule(Main plugin, APIImplementation api) {
		super(ModuleType.Logging, plugin, api);
	}

	@Override
	protected boolean init() {
		return true;
	}

	@Override
	protected void stopIntern() {
		for (LogWrapper wrapper: logWrappers) wrapper.end();
	}
	
	private LogWrapper getWrapper(SubPlugin sub) {
		for (LogWrapper w: logWrappers) if (w.sub == sub) return w;
		
		return null;
	}
	
	public boolean registerSubPlugin(@NotNull SubPlugin sub, @Nullable ModuleStopFunction func) {
		LogWrapper wrapper = new LogWrapper(plugin, sub);
		boolean result = wrapper.init();
		
		if (result) {
			logWrappers.add(wrapper);
			addIntern(sub, func);
		}
		
		return result;
	}
	
	public void unregisterSubPlugin(@Nullable SubPlugin sub) {
		if (isSubPluginLoaded(sub)) {
			LogWrapper wrapper = getWrapper(sub);
			if (wrapper != null) { //Sollte nicht passieren, aber sicher ist sicher
				wrapper.end();
				logWrappers.remove(wrapper);
				removeIntern(sub);
			}
		}
	}
	
	@Override
	public boolean isSubPluginLoaded(SubPlugin sub) {
		return super.isSubPluginLoaded(sub) && getWrapper(sub) != null; //Es wird auch geprüft, ob der Wrapper geladen ist
	}
}

File: src/main/kotlin/io/github/mh321productions/serverapi/module/log/LogModule.kt
package io.github.mh321productions.serverapi.module.log

import org.bukkit.plugin.Plugin
import io.github.mh321productions.serverapi.Main
import io.github.mh321productions.serverapi.api.APIImplementation
import io.github.mh321productions.serverapi.api.SubPlugin
import io.github.mh321productions.serverapi.module.Module
import io.github.mh321productions.serverapi.module.ModuleType
import java.io.IOException

/**
 * Das Logging-Modul erstellt für jedes registrierte Sub-Plugin ein eigenes
 * Logfile, welches automatisch über den internen Logger
 * ([Plugin.getLogger]) beschrieben wird. Wenn das Modul gestoppt
 * oder das Sub-Plugin entladen wird, wird das Logfile geschlossen.
 * Das Logfile ist unter `plugins/ServerAPI/logs/<Sub-Plugin Name>.txt`
 * zu finden.
 * @author 321Productions
 */
class LogModule(plugin: Main, api: APIImplementation) : Module(ModuleType.Logging, plugin, api) {

    private val logWrappers = mutableListOf<LogWrapper>()

    override fun initIntern() = true

    override fun stopIntern() = logWrappers.forEach { it.end() }

    private fun getWrapper(sub: SubPlugin) = logWrappers.firstOrNull { it.sub == sub }

    override fun registerSubPlugin(sub: SubPlugin, func: () -> Unit): Boolean {
        return try {
            val wrapper = LogWrapper(plugin, sub)
            logWrappers.add(wrapper)
            addIntern(sub, func)
            true
        } catch (_: IOException) {
            false
        }
    }

    override fun unregisterSubPlugin(sub: SubPlugin) {
        if (isSubPluginLoaded(sub)) {
            val wrapper = getWrapper(sub)
            if (wrapper != null) { //Sollte nicht passieren, aber sicher ist sicher
                wrapper.end()
                logWrappers.remove(wrapper)
                removeIntern(sub)
            }
        }
    }

    override fun isSubPluginLoaded(sub: SubPlugin): Boolean {
        return super.isSubPluginLoaded(sub) && getWrapper(sub) != null //Es wird auch geprüft, ob der Wrapper geladen ist
    }
}
